src/aspire/reconstruction/mean.py

```diff
@@ -1,17 +1,262 @@
 import logging
+from functools import partial
 
 import numpy as np
+from scipy.linalg import norm
+from scipy.sparse.linalg import LinearOperator, cg
 
+from aspire import config
 from aspire.nufft import anufft
 from aspire.numeric import fft
 from aspire.operators import evaluate_src_filters_on_grid
-from aspire.reconstruction import Estimator, FourierKernel
-from aspire.volume import rotated_grids
+from aspire.reconstruction import Estimator, FourierKernel, FourierKernelMat
+from aspire.volume import Volume, rotated_grids
 
 logger = logging.getLogger(__name__)
 
 
-class MeanEstimator(Estimator):
+class WeightedVolumesEstimator(Estimator):
+    def __init__(self, weights, *args, **kwargs):
+        """
+        Weighted mean volume estimation.
+
+        This is best considered as an r x r matrix of volumes;
+        each volume is a weighted mean least-squares estimator kernel (MeanEstimator).
+        Convolution with each of these kernels is equivalent
+        to performing a projection/backprojection on a volume,
+        with the appropriate amplitude modifiers and CTF,
+        and also a weighting term;
+        the r^2 volumes are each of pairwise products between the weighting vectors given by the columns of wts.
+
+        Note that this is a non-centered Fourier transform, so the zero frequency is found at index 0.
+
+        Formulas and conventions used for Volume estimation are described in:
+
+        “Structural Variability from Noisy Tomographic Projections.”
+        Andén, Joakim, and Amit Singer.
+        SIAM journal on imaging sciences vol. 11,2 (2018): 1441-1492.
+        doi:10.1137/17M1153509
+
+        "Cryo-EM reconstruction of continuous heterogeneity by Laplacian spectral volumes"
+        Amit Moscovich, Amit Halevi, Joakim Andén and Amit Singer
+        Inverse Problems, Volume 36, Number 2, 2020 IOP Publishing Ltd
+        Special Issue on Cryo-Electron Microscopy and Inverse Problems
+        https://doi.org/10.1088/1361-6420/ab4f55
+
+        :param weights: Matrix of weights, n x r.
+        """
+
+        self.weights = weights
+        self.r = self.weights.shape[1]
+        super().__init__(*args, **kwargs)
+        assert self.src.n == self.weights.shape[0]
+
+    def __getattr__(self, name):
+        if name == "precond_kernel":
+            if self.preconditioner == "circulant":
+                # TODO: Discuss precond plans.
+                # self.precond_kernel = FourierKernelMat(
```

Comment:
This doesn't work?

Reply:
The code might be okay, but I think the test doesn't work.... looking now.

Reply:
Yep, just needed to fix broadcast in unit test.  Sorry about that.

Reply:
Cool!